### PR TITLE
statesync: clean up reactor/syncer lifecylce

### PR DESCRIPTION
**Pull request by tychoish**
I've been noticing that there are a number of situations where the
statesync reactor blocks waiting for peers (or similar,) I've moved
things around to improve outcomes in local tests.

### CODE REVIEW

**Comment by tychoish** [comment]
I've been running the `TestReactor_StateProviderP2P` test on repeat locally (10k times), and this seems like it's passing pretty consistently. CI might have different timing constraints though.

File: internal/statesync/reactor.go [hunk 1]
```diff
@@ -288,6 +291,7 @@ func (r *Reactor) Sync(ctx context.Context) (sm.State, error) {
 		r.stateProvider,
 		r.snapshotCh.Out,
 		r.chunkCh.Out,
+		r.snapshotCh.Done(),
```
Comment (williambanfield) [inline internal/statesync/reactor.go:294]: It looks like this is closed when the reactor's closeCh is called.
Any reason to not just rely on this behavior directly? 
https://github.com/tychoish/tendermint/blob/3dd0e3e125bfdf0b9acf7ceed04d070e2094439d/internal/statesync/reactor.go#L872

File: internal/statesync/stateprovider.go [hunk 2]
```diff
@@ -337,44 +337,57 @@ func (s *stateProviderP2P) addProvider(p lightprovider.Provider) {
...
+// (with backoff) until the context is canceled.
 func (s *stateProviderP2P) consensusParams(ctx context.Context, height int64) (types.ConsensusParams, error) {
-	for _, provider := range s.lc.Witnesses() {
-		p, ok := provider.(*BlockProvider)
-		if !ok {
-			panic("expected p2p state provider to use p2p block providers")
-		}
+	iterCount := 0
+	for {
+		for _, provider := range s.lc.Witnesses() {
```
Comment (williambanfield) [inline internal/statesync/stateprovider.go:349]: This function handles a lot. I see what it's doing, but I think the legibility may be helped by putting this inner loop into a helper function that tries every witness and the caller can just orchestrate the retry backoff.

File: internal/statesync/reactor.go (hunk 1 above)
Comment (williambanfield) [reply internal/statesync/reactor.go:294]: Or, can we just pass through the entire `snapshotCh`?

Comment (tychoish) [reply internal/statesync/reactor.go:294]: I tried a few rounds here, but the snapshot channel's lifecycle is managed by the router, not the reactor so using the reactors close channel results in a (possible, though unlikely) send into a closed channel.)

File: internal/statesync/stateprovider.go (hunk 2 above)
Comment (tychoish) [reply internal/statesync/stateprovider.go]: sure! done!

**Review by williambanfield — APPROVED** [review]

**Pull request merged**